### PR TITLE
Monolith variants

### PR DESCRIPTION
**Pull request by n-osborne**
In this PR, I add support for variants in the frontend for Monolith.

### CODE REVIEW

File: src/monolith/printers.ml [hunk 1]
```diff
@@ -7,31 +7,57 @@ let loc = Location.none
 
 let rec ty2printer (ty : Ttypes.ty) =
   match ty.ty_node with
-  | Tyvar tvs -> tvs2printer tvs
-  | Tyapp (tys, _tyl) -> tys2printer tys
+  (* 'a is generated as int but the type checker doesn't know it *)
+  | Tyvar _tvs -> [%expr fun _ -> Print.int 42]
```
[the remark this answers is not in the document]
Comment (n-osborne) [reply src/monolith/printers.ml:11]: The type checker wait for an `'a printer`, so `Print.int` won't be accepted (generates code that does not type check).
It's true that printing `42` is not ideal. Maybe printing something like `"alpha value"` ?
```suggestion
  | Tyvar _tvs -> [%expr fun _ -> Print.string "alhpa value"]
```

File: src/monolith/printers.ml [hunk 2]
```diff
@@ -7,31 +7,57 @@ let loc = Location.none
...
-  | "bool" -> [%expr PPrintOCaml.bool]
-  | "char" -> [%expr PPrintOCaml.char]
-  | "int" -> [%expr PPrintOCaml.int]
-  | "string" -> [%expr PPrintOCaml.string]
-  | s -> failwith (Printf.sprintf "%s printer is not yet implementer" s)
-
-and tys2printer (tys : Ttypes.tysymbol) =
+and tyapp2printer (tys : Ttypes.tysymbol) (tyl : Ttypes.ty list) =
+  let aux i = ty2printer (List.nth tyl i) in
   match tys.ts_ident.id_str with
```
[the remark this answers is not in the document]
Comment (n-osborne) [reply src/monolith/printers.ml:16]: There is no `int` nor `array` in the predefined symbols. Does it make sens to add them?

File: src/monolith/printers.ml [hunk 3]
```diff
@@ -51,10 +77,30 @@ let record_printer (rec_decl : Tast.rec_declaration) =
...
+  let variant (cd : Tast.constructor_decl) =
+    let cname = cd.cd_cs.ls_name.id_str in
+    let cargs = cd.cd_cs.ls_args in
+    let xs =
+      List.init (List.length cargs) (fun _ -> gen_symbol ~prefix:"__x" ())
+    in
+    let xs_str = String.concat ", " xs in
+    let lhs = Printf.sprintf "R.%s (%s)" cname xs_str |> B.pvar in
+    let args = List.map2 ty2repr xs cd.cd_cs.ls_args |> B.elist in
+    let rhs = [%expr PPrintOCaml.variant "" [%e B.estring cname] 0 [%e args]] in
```
[the remark this answers is not in the document]
Comment (n-osborne) [reply src/monolith/printers.ml:94]: Unfortunately, `monolith` does not export a `Print.variant` (neither a `Print.record` for that matter).

File: src/monolith/printers.ml [hunk 4]
```diff
@@ -7,31 +7,57 @@ let loc = Location.none
...
+  | "list" -> [%expr Print.list [%e aux 0]]
+  | "array" -> [%expr Print.array [%e aux 0]]
+  | s when String.sub s 0 5 = "tuple" -> tuple tyl
+  | s ->
+      failwith
+        (Printf.sprintf "%s printer is not yet implemented from tys2printer" s)
+
+and tuple tyl =
+  let elts = List.map (fun _ -> gen_symbol ~prefix:"__t" ()) tyl in
+  let vars = List.map B.evar elts in
```
[the remark this answers is not in the document]
Comment (n-osborne) [reply src/monolith/printers.ml:31]: I need `elts` to make variable expressions and the pattern.

File: src/monolith/printers.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (pascutto) [reply src/monolith/printers.ml:16]: Indeed, they are part of the stdlib, not the built-in symbols, you'll have to get them from there.
I quickly wrote [this](https://github.com/ocaml-gospel/ortac/blob/main/src/core/drv.mli#L7) to do it (interface may change at some point)

File: src/monolith/printers.ml (hunk 4 above)
[the remark this answers is not in the document]
Comment (pascutto) [reply src/monolith/printers.ml:31]: Ah indeed sorry.
Something like that maybe then? (not important, feel free to ignore that)
```suggestion
  let evars, pvars = List.map (fun _ -> let s = gen_symbol ~prefix:"__t" () in B.evar s, B.pvar s) tyl in
  let pat = B.ppat_tuple pvars
```

File: src/monolith/printers.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (pascutto) [reply src/monolith/printers.ml:94]: Can you expose it in your runtime, and use that instead?

**Pull request merged**